packages/itwinui-react/src/core/SearchBox/SearchBox.tsx

```diff
@@ -0,0 +1,398 @@
+  PolymorphicComponentProps,
+  InputFlexContainerProps,
+} from '../utils';
+import { IconButton } from '../Buttons/IconButton';
+import type { IconButtonProps } from '../Buttons/IconButton';
+import '@itwin/itwinui-css/css/input.css';
+
+const SearchBoxContext = React.createContext<
+  | {
+      /**
+       * Context prop for sizing subcomponents
+       */
+      size?: 'small' | 'large';
+      /**
+       * Context prop for disabling subcomponents
+       */
+      disabled?: boolean;
+      /**
+       * Id to pass to input
+       */
+      inputId: string;
+      setInputId: (inputId: string) => void;
+      /**
+       * Ref for input subcomponent
+       */
+      inputRef: React.RefObject<HTMLInputElement>;
+      /**
+       * Ref for open button subcomponent
+       */
+      openButtonRef: React.RefObject<HTMLButtonElement>;
+      /**
+       * Callback for closing expandable searchbox
+       */
+      onClose?: () => void;
+      /**
+       * Callback for expanding searchbox
+       */
+      onOpen?: () => void;
+    }
+  | undefined
+>(undefined);
+
+// ----------------------------------------------------------------------------
+
+type SearchBoxOwnProps = {
+  /**
+   * Whether the searchbox is expandable.
+   * @default false
+   */
+  expandable?: boolean;
+  /**
+   * Searchbox state toggle.
+   */
+  isExpanded?: boolean;
+  /**
+   * Function that is called on toggling (expand, collapse) searchbox.
+   */
+  onToggle?: (isExpanding: boolean) => void;
+  /**
+   * Collapsed searchbox state.
+   * @default SearchBox.OpenButton
+   */
+  collapsedState?: React.ReactNode;
+  /**
+   * Input props when using default searchbox.
+   */
+  inputProps?: React.ComponentPropsWithoutRef<'input'>;
+  /**
+   * Open button props when using default searchbox.
+   */
+  openButtonProps?: React.ComponentPropsWithoutRef<'button'>;
+  /**
+   * Close button props when using default searchbox.
+   */
+  closeButtonProps?: React.ComponentPropsWithoutRef<'button'>;
+  /**
+   * Modify size of the searchbox and it's subcomponents.
+   */
+  size?: 'small' | 'large';
+};
+
+const SearchBoxComponent = React.forwardRef((props, ref) => {
+  useTheme();
+
+  const {
+    size,
+    expandable = false,
+    disabled = false,
+    onToggle,
+    collapsedState,
+    isExpanded,
+    children,
+    inputProps,
+    openButtonProps,
+    closeButtonProps,
+    className,
+    ...rest
+  } = props;
+
+  const [inputId, setInputId] = React.useState(useId());
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const openButtonRef = React.useRef<HTMLButtonElement>(null);
+
+  const [localExpanded, setLocalExpanded] = React.useState(isExpanded);
+
+  const onClose = () => {
+    setLocalExpanded(false);
+    onToggle?.(false);
+    openButtonRef.current && openButtonRef.current.focus();
+  };
+
+  const onOpen = () => {
+    setLocalExpanded(true);
+    onToggle?.(true);
+    inputRef.current && inputRef.current.focus();
+  };
```

Comment:
focus can be fixed using `queueMicrotask`
```suggestion
  const onClose = () => {
    setLocalExpanded(false);
    onToggle?.(false);
    queueMicrotask(() => openButtonRef.current?.focus());
  };

  const onOpen = () => {
    setLocalExpanded(true);
    onToggle?.(true);
    queueMicrotask(() => inputRef.current?.focus());
  };
```

Reply:
Using queueMicrotask! That worked really well. Thanks!